cdap-common/src/main/java/io/cdap/cdap/common/security/AuthEnforceRewriter.java

```diff
@@ -558,14 +560,14 @@ private static int getPreferredParameter(ParameterDetail previousDetails, int pr
           entities = (List<String>) values.get(i + 1);
         } else if ("enforceOn".equals(name)) {
           enforceOn = (Type) values.get(i + 1);
-        } else if ("actions".equals(name)) {
-          // Actions is a ArrayList<String[]> where each String array is of size 2. String[0] is class name Action
-          // class and String[1] is the action which was specified in the annotation
-          List<String[]> actionWithClassName = (List<String[]>) values.get(i + 1);
-          for (String[] action : actionWithClassName) {
-            Preconditions.checkArgument(action.length == 2);
-            Preconditions.checkArgument(Type.getType(action[0]).equals(Type.getType(Action.class)));
-            actions.add(Action.valueOf(action[1]));
+        } else if ("permissions".equals(name)) {
+          // Permissions is a ArrayList<String[]> where each String array is of size 2. String[0] is class name
+          // StandardPermission class and String[1] is the permission which was specified in the annotation
+          List<String[]> permissionWithClassName = (List<String[]>) values.get(i + 1);
+          for (String[] permission : permissionWithClassName) {
+            Preconditions.checkArgument(permission.length == 2);
+            Preconditions.checkArgument(Type.getType(permission[0]).equals(Type.getType(StandardPermission.class)));
```

Comment:
Yes. It's not possible to have an interface parameter in annotation. There are two options here: string constants or just limit now to StandardPermission

Reply:
Actually one more option would be to use nested annotations. Let me see if I can make it happen

Reply:
So, it was under 10 usages of this annotation before this PR and a single usage after. I am not sure if it worth spending time on it or may be just remove altogether. The rewriting code is pretty complex and any error in it can be a security breach. If we want to support it, it would need additional annotations to support more permission types, e.g. `@EnforceApplication(ApplicationPermission.PREVIEW)`

Reply:
Sounds good. IIRC there was an original push to eventually migrate everything to use the annotation, but, if we see more enforcement on the handler level instead of the service interface level, this will be an infeasible task anyways.